Update pychop definitions for SNS instruments

mducle opened the pull request:
### Description of work

When SNS instruments were added to PyChop around 2020 I used the original `.yaml` files rather than fine-tuned files which Jiao Lin had created during the work for [this publication](https://www.sciencedirect.com/science/article/pii/S0921452618307233). So, the calculations in Mantid has actually been incorrect for 4 years!

During work for the [resconvlib](https://github.com/pace-neutrons/resolution_functions/) project @RastislavTuranyi found that Mantid calculations disagreed with the online calculator [provided by SNS](https://rez.mcvine.ornl.gov/) which is also meant to be using PyChop. Furthermore the Mantid calculations disagreed drastically with the measured data in the ARCS resolution paper.

The discrepancy is because the moderator parameters (amongst other things) were incorrect in the initial version added to Mantid. 

#### Summary of work
<!-- Please provide a short, high level description of the work that was done.
-->

This PR is to update these `.yaml` files to the optimised files in Jiao Lin's [original repository](https://github.com/sns-chops/resolution), namely the files:

* [ARCS/PyChop/arcs-09122018.yaml](https://github.com/sns-chops/resolution/blob/master/ARCS/PyChop/arcs-09122018.yaml)
* [SEQUOIA/PyChop/sequoia-06212019.yaml](https://github.com/sns-chops/resolution/blob/master/SEQUOIA/PyChop/sequoia-06212019.yaml)

<!-- Why has this work been done? If there is no linked issue please provide appropriate context for this work.
#### Purpose of work
This can be removed if a github issue is referenced below
-->
<!-- alternative
Fixes #xxxx. <!-- and fix #xxxx or close #xxxx xor resolves #xxxx. One line per issue fixed. -->

*There is no associated issue.*

<!-- If the original issue was raised by a user they should be named here. Do not leak email addresses
**Report to:** [user name]
-->

#### Further detail of work
<!-- Please provide a more detailed description of the work that has been undertaken.
-->

### To test:

1. Run Mantid and open the PyChop interface (Direct->PyChop).
2. Select ARCS and type in an incident energy (say 100meV) and click calculate.
3. Go to the [SNS resolution webpage](https://rez.mcvine.ornl.gov/); select ARCS and use the same parameters as in the Mantid/PyChop calculation, and check the calculations agree.
4. Check that the units for flux in all the plots (and plot labels) are `n/cm^2/s/MW`.
5. Repeat for Sequoia.
6. Select MERLIN (or any other ISIS instruments, LET, MARI or MAPS) and click calculate. Check that the flux units are `n/cm^2/s/160uA`.

<!-- Instructions for testing.
There should be sufficient instructions for someone unfamiliar with the application to test - unless a specific
reviewer is requested.
If instructions for replicating the fault are contained in the linked issue then it is OK to refer back to these.
-->

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
If you add release notes please save them as a separate file using the Issue or PR number as the file name. Check the file is located in the correct directory for your note(s).
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

mducle committed:
Update pychop definitions for SNS instruments

SilkeSchomann commented:
@mducle Please add a release note.

ajjackson commented:
I notice that in the .yaml for SEQUOIA, the comment on moderator parameters has not changed and indicates that B will be used directly (rather than as reciprocal).

I compared these comments a bit more closely with the code, and _neither_ of them seems to be an accurate description of the `tikeda` function https://github.com/mantidproject/mantid/blob/6ee0b15d7d58be39fc02daf7c8d01a9f384471de/scripts/pychop/Chop.py#L116-L129

Specifically:
- The final expression is effectively $\tau^2 = 3 \tau_f^2 + R(2-R)\tau_s^2$; the components of $\tau$ are _not_ reciprocals. (This makes sense as it means all the $\tau_i$ have the same units.)
- I can't figure out where the conversion 81.8048 comes from. It should be converting an energy in meV to a wavelength according to the Ikeda-Carpenter paper.
- The $R(2-R)$ group in the implementation looks a bit odd, and differs from the comments which describe $R/(2-R)$. Neither form obviously resembles the Ikeda-Carpenter function in the literature; I think there must be some algebraic simplification at work?

refs:
- [Mantid docs](https://docs.mantidproject.org/v6.1.0/fitting/fitfunctions/IkedaCarpenterPV.html)
- [Ikeda-Carpenter paper](https://www.sciencedirect.com/science/article/pii/0168900285900336)
- [Paper on fitting Ikeda-Carpenter functions for instruments](https://www.epj-conferences.org/articles/epjconf/pdf/2015/02/epjconf-Qens-Wins2014_03001.pdf)

mducle commented:
`81.81` is a ["well-known" conversion factor](https://books.google.co.uk/books?id=Lx4xcz3v9IMC&pg=PP3&source=gbs_selected_pages&cad=1#v=snippet&q=%22velocity%20energy%20and%20temperature%20for%20thermal%20neutrons%22&f=false) for neutron conversion where `E = 81.81 / lambda**2` when energy is in `meV` and wavelength is in Angstrom. It amounts to `h/2m`.

I'll see if I can dig out where the approximation comes from... it's probably more fully explained in the original Fortran code, which I have... somewhere...

Also the unit test is broken because it compares numerical values of the resolution and flux which is obviously now different (but correct!). I'll fix that and add release notes.

AndreiSavici commented:
@ajjackson The 81.81 is the conversion from energy in meV to wavelength in Angstroms. $h^2/(2m_n\lambda^2)=\Delta E$ But this formula is in SI. We use $\lambda$ in Angstroms, and $\Delta E (meV)=\Delta E(J) 10^{-3}e$, where $e$ is the elementary charge. So the constant is 

$$\frac{h^2}{2m_n e (10^{10})^210^{-3}}$$

mducle committed:
PyChop-SNS: Add docs. Got agreement with rez.mcvine.ornl.gov

mducle commented:
@ajjackson The original Fortran code is in a [gist here](https://gist.github.com/mducle/de862bf6e637f8f80df08a6ddf3f9208). Not sure it's super useful. I think that the value returned by the function, `tausqr` is variance of the distribution in eq 10' of Ikeda-Carpenter so it should be $\int \psi^2(v, t) dt$... I might see if I can plug that into Wolfram...

mducle commented:
@AndreiSavici I've changed the `.yaml` files here a bit compared to the [arcs-09122018.yaml](https://github.com/sns-chops/resolution/blob/master/ARCS/PyChop/arcs-09122018.yaml) and [sequoia-06212019.yaml](https://github.com/sns-chops/resolution/blob/master/SEQUOIA/PyChop/sequoia-06212019.yaml) files in order to match what the [website](https://rez.mcvine.ornl.gov/) gives instead. 

For ARCS I've included a [scale factor of 600](https://github.com/sns-chops/resolution/blob/master/dashui/arcs/model.py#L12) for the flux and for SEQUOIA I've changed the look up table to include the [detector efficiency and absorption corrections](https://github.com/sns-chops/resolution/blob/master/dashui/sequoia/model.py#L20) which the website code does. 

So now Mantid-PyChop will give the same flux (in n/cm^2/s/MW) as the webpage.

@SilkeSchomann I've also changed the documentation to explicitly say that the flux values given for ISIS and SNS instruments are in different units. (Although in principle if both facilities are running at full power the flux should be in the same units of n/cm^2/s). I haven't changed the GUI - it still says n/cm^2/s - do you think that's needed for this?

SilkeSchomann commented:
> @AndreiSavici I've changed the `.yaml` files here a bit compared to the [arcs-09122018.yaml](https://github.com/sns-chops/resolution/blob/master/ARCS/PyChop/arcs-09122018.yaml) and [sequoia-06212019.yaml](https://github.com/sns-chops/resolution/blob/master/SEQUOIA/PyChop/sequoia-06212019.yaml) files in order to match what the [website](https://rez.mcvine.ornl.gov/) gives instead.
> 
> For ARCS I've included a [scale factor of 600](https://github.com/sns-chops/resolution/blob/master/dashui/arcs/model.py#L12) for the flux and for SEQUOIA I've changed the look up table to include the [detector efficiency and absorption corrections](https://github.com/sns-chops/resolution/blob/master/dashui/sequoia/model.py#L20) which the website code does.
> 
> So now Mantid-PyChop will give the same flux (in n/cm^2/s/MW) as the webpage.
> 
> @SilkeSchomann I've also changed the documentation to explicitly say that the flux values given for ISIS and SNS instruments are in different units. (Although in principle if both facilities are running at full power the flux should be in the same units of n/cm^2/s). I haven't changed the GUI - it still says n/cm^2/s - do you think that's needed for this?

@mducle I would change the GUI as well. A mismatch between GUI and documentation will probably confuse the user.

ajjackson commented on scripts/pychop/arcs.yaml:
```diff
@@ -106,15 +106,12 @@ moderator:
-             .037, .17, 172.42] #    where tau_f=1/(sqrt(Ei)*E2V*sqrt(S1^2+(S2*lam)^2), tau_s=B1 for E<=130meV, B2 otherwise, R=exp(-Ei/Emod)
+  mod_pars: [281., 79.0,        # Parameters for time profile (for I-K is: [S1, S2, B1, B2, Emod]
+             .087, .4, 172.]    #    where tau_f=1/(sqrt(Ei)*E2V*sqrt(S1^2+(S2*lam)^2), tau_s=1/B1 for E<=130meV, 1/B2 otherwise, R=exp(-Ei/Emod)
+                                #    and the overall variance (not FWHM) is: tau^2 = 3/tau_f^2 + R/(2-R)/tau_s^2
```
```suggestion
                                #    and the overall variance (not FWHM) is: tau^2 = 3 tau_f^2 + R(2-R) tau_s^2
```

I think this is more accurate? May need updating in other files if so.

mducle committed:
PyChop-SNS: Add flux units to GUI and yaml

mducle committed:
PyChop-SNS: correct Ikeda-Carpenter eqn in yaml

mducle committed:
PyChop-SNS: Fix bug wrong freq for aperture width

mducle commented:
@AndreiSavici Sorry about this. There was a bug in the [code](https://github.com/mantidproject/mantid/blob/e4575ed635f1d0ecbb5b228719d8372ffe7569fc/scripts/pychop/Instruments.py#L891-L898) to calculate the aperture time width. The is the time uncertainty due to the distance uncertainties caused by the finite size of the moderator face as seen from the sample. The code takes into account how the chopper will reduce the effective width as seen from the sample (and thus reduce the time width) when it rotates faster and so requires the chopper frequency. However in the code I took this frequency from the `Instrument` object rather than from the `Chopper` object where it is actually set. In testing I was using the `calculate` method which sets both but in the GUI only the frequency in the `Chopper` is set and the frequency in the `Instrument` object is left at the default (300Hz for ARCS). So PyChop was always calculating a larger aperture width then it should.

For most instruments this doesn't really matter (the aperture width is normally only ~1-2% of the overall time width), but it seems it _is_ important for ARCS, and I think this is because in the [ARCS definition](https://github.com/mantidproject/mantid/blob/e4575ed635f1d0ecbb5b228719d8372ffe7569fc/scripts/pychop/arcs.yaml#L8-L9) the aperture is defined to be 175x195mm whereas for most other instruments (SEQUOIA, MAPS, MERLIN) it is around 50x50 or 60x60mm.

SilkeSchomann reviewed (approved):
I've checked both the plots and the data output window for ARCS and Sequoia for an incident energy of 100meV and a frequency of 300. The results matched the corresponding results on the SNS resolution webpage. The units for flux on the plots and in the data window were n/cm^2/s/MW. 
For MERLIN, I have compared the plots and the data output window. The units for flux were always n/cm^2/s/160uA.

SilkeSchomann commented:
@mantidproject/gatekeepers  It would be good to wait for the review from @AndreiSavici  as well as I don't know for which chopper/energy/frequency combinations for ARCS he found mismatches.

AndreiSavici commented:
Looks good
